src/runtime_src/core/pcie/tools/cloud-daemon/msd.cpp

```diff
@@ -286,26 +288,39 @@ static void msd(std::shared_ptr<pcidev::pci_device> d, std::string host)
             }
         }
 
+        retfd[0] = retfd[1] = -100;
         // Waiting for msg to show up, interval is 3 seconds.
-        ret = waitForMsg(dev, mbxfd, mpdfd, 3);
+        ret = waitForMsg(dev, mbxfd, mpdfd, 3, retfd);
         if (ret < 0) {
             if (ret == -EAGAIN) // MPD has been quiet, retry.
                 continue;
             else
                 break;
         }
 
+        msg.localFd = mbxfd;
+        msg.remoteFd = mpdfd;
+
         // Process msg.
-        if (ret == mbxfd)
-            ret = processLocalMsg(dev, mbxfd, mpdfd);
-        else
-            ret = processRemoteMsg(dev, mbxfd, mpdfd, remoteMsgHandler);
-        if (ret == -EAGAIN) { // Socket connection was lost, retry
-            close(mpdfd);
-            mpdfd = -1;
-            continue;
-        } else if (ret != 0) {
-            break;
+        for (int i = 0; i < 2; i++) {
+            if (retfd[i] == mbxfd) {
+                msg.type = LOCAL_MSG;
+                msg.data = getLocalMsg(dev, mbxfd);
+            } else if (retfd[i] == mpdfd) {
+                msg.type = REMOTE_MSG;
+                msg.cb = remoteMsgHandler;
+            } else
+                continue;
+            if (msg.data == nullptr)
+                goto done;
```

Comment:
goto have no place in c++.  manage the resources with a smart pointer and return.

Reply:
I would be less strong... In some case `goto` are useful in C++ too... :-)

Reply:
remove the goto

Reply:
@keryell Very rarely useful, but definitely not for "do_something; clean up",  where RAII should be used.